sqlglot/optimizer/qualify_columns.py

```diff
@@ -422,10 +459,86 @@ def _expand_stars(
             tables = [expression.table]
             _add_except_columns(expression.this, tables, except_columns)
             _add_replace_columns(expression.this, tables, replace_columns)
+
+        elif expression.is_star and (
+            isinstance(expression, exp.Dot)
+            or (isinstance(expression, exp.Column) and expression.table in scope.sources)
+        ):
+            col = expression.find(exp.Column) if isinstance(expression, exp.Dot) else expression
+            starting_struct = str(expression.this) if isinstance(expression.this, exp.Dot) else None
+            structs.append((col, starting_struct))
+
+            # Annotate scope ahead of the struct flattening
+            if not scope_annotated and dialect == "bigquery":
+                annotator.annotate_scope(scope)
+                scope_annotated = True
```

Comment:
Can we use another signal to conditionally annotate here instead of adding this flag-controlled branch inside the loop? Perhaps check if there's anything in `scope.stars` and if it's a dot, then ...

Reply:
Will look into it, bear in mind that you could have more expansions in the same scope as with the added test of 2 tables, so this was the safest condition until now

Reply:
Can you explain this a bit more? Not sure I get why it'd be unsafe.

Reply:
Not unsafe per se, but if the condition has the "risk" that it can be repeated then we might be annotating the scope over and over again since we're under a loop